src/Orleans/Providers/IEventStorageProvider.cs

```diff
@@ -0,0 +1,117 @@
+using System.Threading.Tasks;
+using Orleans.Providers;
+using Orleans.Runtime;
+using System.Collections.Generic;
+
+namespace Orleans.Storage
+{
+    /// <summary>
+    /// Interface to be implemented for a event storage provider able to read from and append to streams
+    /// representing the state of Journaled Grains
+    /// </summary>
+    public interface IEventStorageProvider : IProvider, IEventStorage
+    {
+        /// <summary>Logger used by this event storage provider instance.</summary>
+        /// <returns>Reference to the Logger object used by this provider.</returns>
+        /// <seealso cref="Logger"/>
+        Logger Log { get; }
+
+        /// <summary>
+        /// The provider can specify a default stream name to use for a given grain.
+        /// </summary>
+        /// <param name="grainType">The type of the grain</param>
+        /// <param name="grainReference">The grain reference for the grain</param>
+        /// <returns></returns>
+        string DefaultStreamName(Type grainType, GrainReference grainReference);
+    }
+
+    /// <summary>
+    /// The event store interface, as implemented by event storage providers
+    /// </summary>
+    /// 
+    public interface IEventStorage
+    {
+        /// <summary>
+        /// Returns a handle to an event stream that can be used to perform operations on it.
+        /// </summary>
+        /// <param name="streamName">The name of the stream</param>
+        /// <returns>A handle for performing operations on the stream</returns>
+        IEventStreamHandle GetEventStreamHandle(string streamName);
+    }
+
+    /// <summary>
+    /// The event stream interface, as implemented by event storage providers. 
+    /// Should be disposed when no longer needed.
+    /// </summary>
+    public interface IEventStreamHandle : IDisposable
+    {
+        /// <summary>
+        /// The name of this event stream.
+        /// </summary>
+        string StreamName { get; }
+
+        /// <summary>
+        /// Returns the number of events in the event stream, which corresponds to the version number.
+        /// Returns zero for a not-yet-created stream or for a deleted stream.
+        /// </summary>
+        /// <returns>a nonnegative integer representing the number of events in the stream.</returns>
+        Task<int> GetVersion();
+
+        /// <summary>
+        /// Loads a sequence of events from the event stream.
+        /// </summary>
+        /// <typeparam name="E">The base class for the events</typeparam>
+        /// <param name="startAtVersion">The version (stream position) at which to start</param>
+        /// <param name="endAtVersion">The version (stream position) at which to end, or null to return up and including the most recent</param>
+        /// <returns>A <see cref="EventStreamSegment{E}"/> structure containing a subrange of the event stream.</returns>
+        Task<EventStreamSegment<E>> Load<E>(int startAtVersion = 0, int? endAtVersion = null);
```

Comment:
As I mentioned in other comments, but they might be lost as they are stale, I don't think making all the methods in IEventStreamHandle was the right choice. This is not required by the implementation, and seems to be just syntactic sugar for the current usage pattern, but not necessarily even by design, just a side effect of how you generated the collections before calling into these methods, but it could have been different too